cpp/src/arrow/builder.cc

```diff
@@ -1413,6 +1413,41 @@ Status StringBuilder::Append(const std::vector<std::string>& values,
   return Status::OK();
 }
 
+Status StringBuilder::Append(const char** values,
+                             int64_t length,
+                             const uint8_t* valid_bytes) {
+  std::size_t total_length = 0;
+  std::vector<std::size_t> value_lengths(length);
+  for (int64_t i = 0; i < length; ++i) {
+    if (values[i]) {
+      auto value_length = strlen(values[i]);
+      value_lengths[i] = value_length;
+      total_length += value_length;
+    }
+  }
+  RETURN_NOT_OK(Reserve(length));
+  RETURN_NOT_OK(value_data_builder_.Reserve(total_length));
+  RETURN_NOT_OK(offsets_builder_.Reserve(length));
+
+  if (valid_bytes) {
+    for (int64_t i = 0; i < length; ++i) {
+      RETURN_NOT_OK(AppendNextOffset());
+      if (valid_bytes[i]) {
+        RETURN_NOT_OK(value_data_builder_.Append(
+            reinterpret_cast<const uint8_t*>(values[i]), value_lengths[i]));
+      }
+    }
+  } else {
+    for (int64_t i = 0; i < length; ++i) {
+      RETURN_NOT_OK(AppendNextOffset());
+      RETURN_NOT_OK(value_data_builder_.Append(
+          reinterpret_cast<const uint8_t*>(values[i]), value_lengths[i]));
```

Comment:
Does this create a non-null array item even if `values[i]` is NULL?

Reply:
It's interesting.
I've implemented as the followings:

  * If `values[i]` is `NULL` and `valid_bytes` isn't `nullptr`, the value is an empty string not a null value. It's for respecting `valid_bytes` data.
  * If `values[i]` is `NULL` and `valid_bytes` is `nullptr`, the value is a null value.

But users may be confused...

Reply:
Hmm, yes, I think it's a bad idea. If we accept `NULL` value pointers, it should IMHO *always* mean a null item in the array (regardless of `valid_bytes`). If we don't accept `NULL` value pointers, we should at least add a DCHECK against them.

I think it's a good idea to accept them, that way people don't have to allocate a separate `valid_bytes` array.